google-cloud-firestore/src/main/java/com/google/cloud/firestore/RecursiveDelete.java

```diff
@@ -65,27 +65,41 @@
 
   private final FirestoreRpcContext<?> firestoreRpcContext;
   private final BulkWriter writer;
-  @Nullable private final CollectionReference collectionReference;
+
   @Nullable private final DocumentReference documentReference;
 
+  private final ResourcePath parentPath;
+  private final String collectionId;
+
+  /** Lock object for all BulkWriter operations and callbacks. */
+  private final Object lock = new Object();
+
   /** The number of deletes that failed with a permanent error. */
+  @GuardedBy("lock")
   private int errorCount = 0;
 
   /**
    * The most recently thrown error. Used to populate the developer-facing error message when the
    * recursive delete operation completes.
    */
-  @Nullable private Throwable lastError;
+  @GuardedBy("lock")
+  @Nullable
+  private Throwable lastError;
 
   /** Whether there are still documents to delete that still need to be fetched. */
+  @GuardedBy("lock")
   private boolean documentsPending = true;
 
   /** A deferred promise that resolves when the recursive delete operation is completed. */
   private final SettableApiFuture<Void> completionFuture = SettableApiFuture.create();
 
   /** Whether a query stream is currently in progress. Only one stream can be run at a time. */
+  @GuardedBy("lock")
   private boolean streamInProgress = false;
```

Comment:
You might be able to combine `streamInProgress` and `documentsPending` with some `!` magic.